### PR TITLE
Add zone layout and CGMES implementation

### PR DESCRIPTION
**Pull request by massimo-ferraro**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#8

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Comment by massimo-ferraro** [comment]
The PR includes code in single-line-diagram-core and single-line-diagram-cgmes modules.
Classes added to single-line-diagram-core module are partially tested by test classes included in single-line-diagram-cgmes module (e.g. com.powsybl.sld.model.LineEdge class, tested also by com.powsybl.sld.cgmes.layout.CgmesZoneLayoutTest).
SonarCloud seems not to realize it, and it seems to take into account only tests added to the same module to compute the coverage of a class. So the coverage of LineEdge class is 25% (the code tested in a test defined in single-line-diagram-core), but most of the remaining code is tested by CgmesZoneLayoutTest.
The results is that the PR has, so far, 77.2% coverage, below the required 80%.
Is it possible to make SonarCloud use also the tests in other modules to compute the coverage? Or is it required to have all the tests for a module classes in the same module?

**Comment by geofjamg** [comment]
> 
> 
> The PR includes code in single-line-diagram-core and single-line-diagram-cgmes modules.
> Classes added to single-line-diagram-core module are partially tested by test classes included in single-line-diagram-cgmes module (e.g. com.powsybl.sld.model.LineEdge class, tested also by com.powsybl.sld.cgmes.layout.CgmesZoneLayoutTest).
> SonarCloud seems not to realize it, and it seems to take into account only tests added to the same module to compute the coverage of a class. So the coverage of LineEdge class is 25% (the code tested in a test defined in single-line-diagram-core), but most of the remaining code is tested by CgmesZoneLayoutTest.
> The results is that the PR has, so far, 77.2% coverage, below the required 80%.
> Is it possible to make SonarCloud use also the tests in other modules to compute the coverage? Or is it required to have all the tests for a module classes in the same module?

Sonar is not correctly configured to build an aggregated report like it is in powsybl core. I fill an issue for that.

**Comment by geofjamg** [comment]
> 
> 
> > The PR includes code in single-line-diagram-core and single-line-diagram-cgmes modules.
> > Classes added to single-line-diagram-core module are partially tested by test classes included in single-line-diagram-cgmes module (e.g. com.powsybl.sld.model.LineEdge class, tested also by com.powsybl.sld.cgmes.layout.CgmesZoneLayoutTest).
> > SonarCloud seems not to realize it, and it seems to take into account only tests added to the same module to compute the coverage of a class. So the coverage of LineEdge class is 25% (the code tested in a test defined in single-line-diagram-core), but most of the remaining code is tested by CgmesZoneLayoutTest.
> > The results is that the PR has, so far, 77.2% coverage, below the required 80%.
> > Is it possible to make SonarCloud use also the tests in other modules to compute the coverage? Or is it required to have all the tests for a module classes in the same module?
> 
> Sonar is not correctly configured to build an aggregated report like it is in powsybl core. I fill an issue for that. #40

Fixed.

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/ZoneLayoutFactory.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.powsybl.sld.model.ZoneGraph;
+
+/**
+ *
+ * @author Massimo Ferraro <massimo.ferraro@techrain.eu>
+ */
+public interface ZoneLayoutFactory {
+
+    ZoneLayout create(ZoneGraph graph);
```
Comment (geofjamg) [inline single-line-diagram-core/src/main/java/com/powsybl/sld/layout/ZoneLayoutFactory.java]: As for SubstationLayoutFactory we should be able to pass a SubstationLayoutFactory and a  VoltageLayoutFactory to recursively layout a zone with only automatic layouts.

**Pull request merged**